### DISCUSSION TITLE
[QUESTION] Understand file retrieving long duration

### DISCUSSION THREAD

**Discussion by Tom3652** [fluttercandies/flutter_photo_manager#661] [Q&A]
## Expect
I need the ```filePath``` of the ```AssetEntity``` to be able to upload it.
So i have tested : 
- ```assetEntity.file``` 
- ```assetEntity.originFile```
- ```assetEntity.loadFile()```

I am also trying to display very quickly a preview of the file (thanks to its path) in a ```VideoPlayer``` Widget.
However, my beta test users told me that videos were not playing locally for a very long time and after it's always working.

So i have set a ```PMProgressHandler``` to listen to the progress as written in the docs, and it's working. 
However i am surprised to see the progress like this : 

```
flutter: Event : PMRequestState.prepare, progress : 0.0 & time : 47
flutter: Event : PMRequestState.loading, progress : 0.01 & time : 49
flutter: Event : PMRequestState.loading, progress : 0.02 & time : 49
flutter: Event : PMRequestState.loading, progress : 0.03 & time : 50
flutter: Event : PMRequestState.loading, progress : 0.04 & time : 50
flutter: Event : PMRequestState.loading, progress : 0.05 & time : 51
flutter: Event : PMRequestState.loading, progress : 0.06 & time : 51
flutter: Event : PMRequestState.loading, progress : 0.07 & time : 52
flutter: Event : PMRequestState.loading, progress : 0.08 & time : 52
flutter: Event : PMRequestState.loading, progress : 0.09 & time : 52
flutter: Event : PMRequestState.loading, progress : 0.1 & time : 53
flutter: Event : PMRequestState.loading, progress : 0.11 & time : 53
flutter: Event : PMRequestState.loading, progress : 0.12 & time : 54
flutter: Event : PMRequestState.loading, progress : 0.13 & time : 54
flutter: Event : PMRequestState.loading, progress : 0.14 & time : 55
flutter: Event : PMRequestState.loading, progress : 0.15 & time : 55
flutter: Event : PMRequestState.loading, progress : 0.16 & time : 55
flutter: Event : PMRequestState.loading, progress : 0.17 & time : 56
flutter: Event : PMRequestState.loading, progress : 0.18 & time : 56
flutter: Event : PMRequestState.loading, progress : 0.19 & time : 57
flutter: Event : PMRequestState.loading, progress : 0.2 & time : 57
flutter: Event : PMRequestState.loading, progress : 0.21 & time : 58
flutter: Event : PMRequestState.loading, progress : 0.22 & time : 58
flutter: Event : PMRequestState.loading, progress : 0.23 & time : 59
flutter: Event : PMRequestState.loading, progress : 0.24 & time : 59
flutter: buffer video false
flutter: Event : PMRequestState.loading, progress : 0.25 & time : 0
flutter: Event : PMRequestState.loading, progress : 0.26 & time : 1
flutter: Event : PMRequestState.loading, progress : 0.27 & time : 1
flutter: Event : PMRequestState.loading, progress : 0.28 & time : 1
flutter: Event : PMRequestState.loading, progress : 0.3 & time : 2
flutter: Event : PMRequestState.loading, progress : 0.31 & time : 3
flutter: buffer video false
flutter: Event : PMRequestState.loading, progress : 0.32 & time : 4
flutter: Event : PMRequestState.loading, progress : 0.33 & time : 4
flutter: Event : PMRequestState.loading, progress : 0.34 & time : 5
flutter: Event : PMRequestState.loading, progress : 0.35 & time : 5
flutter: Event : PMRequestState.loading, progress : 0.36 & time : 6
flutter: Event : PMRequestState.loading, progress : 0.37 & time : 7
flutter: Event : PMRequestState.loading, progress : 0.38 & time : 8
flutter: Event : PMRequestState.loading, progress : 0.39 & time : 8
flutter: Event : PMRequestState.loading, progress : 0.4 & time : 9
flutter: Event : PMRequestState.loading, progress : 0.41 & time : 9
flutter: buffer video false
flutter: Event : PMRequestState.loading, progress : 0.42 & time : 10
flutter: buffer video false
flutter: Event : PMRequestState.loading, progress : 0.43 & time : 10
flutter: Event : PMRequestState.loading, progress : 0.44 & time : 11
flutter: buffer video false
flutter: Event : PMRequestState.loading, progress : 0.45 & time : 11
flutter: Event : PMRequestState.loading, progress : 0.46 & time : 12
flutter: Event : PMRequestState.loading, progress : 0.47 & time : 12
flutter: Event : PMRequestState.loading, progress : 0.48 & time : 13
flutter: Event : PMRequestState.loading, progress : 0.49 & time : 13
flutter: Event : PMRequestState.loading, progress : 0.5 & time : 14
flutter: Event : PMRequestState.loading, progress : 0.51 & time : 14
flutter: Event : PMRequestState.loading, progress : 0.52 & time : 15
flutter: Event : PMRequestState.loading, progress : 0.53 & time : 16
flutter: Event : PMRequestState.loading, progress : 0.54 & time : 16
flutter: Event : PMRequestState.loading, progress : 0.55 & time : 17
flutter: Event : PMRequestState.loading, progress : 0.56 & time : 17
flutter: Event : PMRequestState.loading, progress : 0.57 & time : 18
flutter: Event : PMRequestState.loading, progress : 0.59 & time : 19
flutter: Event : PMRequestState.loading, progress : 0.6 & time : 19
flutter: Event : PMRequestState.loading, progress : 0.61 & time : 20
flutter: Event : PMRequestState.loading, progress : 0.62 & time : 21
flutter: Event : PMRequestState.loading, progress : 0.63 & time : 21
flutter: Event : PMRequestState.loading, progress : 0.64 & time : 22
flutter: Event : PMRequestState.loading, progress : 0.65 & time : 22
flutter: Event : PMRequestState.loading, progress : 0.66 & time : 23
flutter: Event : PMRequestState.loading, progress : 0.68 & time : 23
flutter: Event : PMRequestState.loading, progress : 0.69 & time : 23
flutter: Event : PMRequestState.loading, progress : 0.7 & time : 24
flutter: Event : PMRequestState.loading, progress : 0.72 & time : 24
flutter: Event : PMRequestState.loading, progress : 0.73 & time : 24
flutter: Event : PMRequestState.loading, progress : 0.74 & time : 25
flutter: Event : PMRequestState.loading, progress : 0.75 & time : 25
flutter: Event : PMRequestState.loading, progress : 0.76 & time : 25
flutter: Event : PMRequestState.loading, progress : 0.78 & time : 25
flutter: Event : PMRequestState.loading, progress : 0.79 & time : 26
flutter: Event : PMRequestState.loading, progress : 0.8 & time : 26
flutter: Event : PMRequestState.loading, progress : 0.81 & time : 26
flutter: Event : PMRequestState.loading, progress : 0.82 & time : 27
flutter: Event : PMRequestState.loading, progress : 0.84 & time : 27
flutter: Event : PMRequestState.loading, progress : 0.85 & time : 27
flutter: Event : PMRequestState.loading, progress : 0.86 & time : 28
flutter: Event : PMRequestState.loading, progress : 0.87 & time : 28
flutter: Event : PMRequestState.loading, progress : 0.88 & time : 28
flutter: Event : PMRequestState.loading, progress : 0.9 & time : 29
flutter: Event : PMRequestState.loading, progress : 0.91 & time : 29
flutter: Event : PMRequestState.loading, progress : 0.92 & time : 30
flutter: Event : PMRequestState.loading, progress : 0.93 & time : 30
flutter: Event : PMRequestState.loading, progress : 0.94 & time : 30
flutter: Event : PMRequestState.loading, progress : 0.95 & time : 31
flutter: Event : PMRequestState.loading, progress : 0.96 & time : 31
flutter: Event : PMRequestState.loading, progress : 0.97 & time : 32
flutter: Event : PMRequestState.loading, progress : 0.98 & time : 32
flutter: Event : PMRequestState.loading, progress : 0.99 & time : 33
flutter: Event : PMRequestState.success, progress : 1.0 & time : 33
flutter: Displayed loaded file is : File: '/private/var/mobile/Containers/Data/Application/C92417AB-642F-4FA2-825B-1F380C2E59C7/tmp/.video/IMG_8696.MOV'
```

Where ```time``` is in ```seconds```. The above logs are given by the below code : 

```
 PMProgressHandler handler = PMProgressHandler();
 handler.stream.listen((event) {
      print("Event : ${event.state}, progress : ${event.progress} & time : ${DateTime.now().second}");
 });
 File? displayedFile = await assetEntity.loadFile(isOrigin: true, progressHandler: handler);
```

So as you can see, it tooks 33 + (60-47) = 46 seconds to load a 19sec video file.
I would like to understand why is it so long to retrieve a path ? This is almost x5 times the time it takes to actually upload the video  :/ 

The only explanation i have came up with is that the path actually doesn't exist and you save the origin file to a temp file path and that takes a huge amount of time even for a powerful device... Which is also weird because i am also saving temp files in my app (1min videos that are edited) and it usually takes maximum 5seconds for the biggest video. 

I need this ```filePath``` to "exists" as fast as possible, is there any way to do that ? 

Thanks a lot in advance for any guidance !

**Comment by Tom3652** [comment]
This is actually more a bug than a question because i can't upload files because of this reason, even on very performant smartphones it takes too long to retrieve...

Is there any way to make it available instantly like getting a `File` in Flutter ?

**Comment by AlexV525** [comment]
If `await entity.isLocallyAvailable` equals `false`, which typically means:
- Your entity is only saved on iCloud storage rather than your device. Then the speed of retrieving the file depends on the speed of download.
- The resource you're requesting is not cached into the sandbox of the app. This is required by the iOS and cannot be avoid, which makes the first I/O fetching become few seconds.

All of these conditions are by design and must be implemented accordingly. If you want to know more about how to implement this, we've implemented a more user-friendly interaction at here: https://github.com/fluttercandies/flutter_wechat_assets_picker/blob/973fdcbb469cb2b2c7fc9750f0adf199f69a64d5/lib/src/widget/builder/locally_available_builder.dart

> Is there any way to make it available instantly like getting a File in Flutter ?

The answer is **NO** in most of the time on iOS. This system has a *Sandbox* mechanism, which will copy newly retrieving files into the sandbox that the app is holding.

**Comment by Tom3652** [comment]
Thanks for the detailed answer !

**Comment by Tom3652** [comment]
So it is possible that the asset comes from even if we have not enable iCloud (i have never put photos / videos in iCloud and i have this loading on some files), i had no idea about it.

How can the thumbnail be retrieved instantly with ```entity.thumbDataWithOption()``` and when calling ```entity.file``` on the same entity it downloads it ? 
In other words, how can we get a thumbnail from a file that is not ```locallyAvailable``` ? Unless all thumbnails are stored on the device and can be retrieved without the original file.

I would guess that Apple make them available for any apps to avoid using too much disk space per app for caching always the same assets but i don't know...

**Comment by AlexV525** [comment]
> So it is possible that the asset comes from even if we have not enable iCloud (i have never put photos / videos in iCloud and i have this loading on some files), i had no idea about it.

Is it a real question? As you can think this again, it's impossible for us to retrieve any files that are not available locally.

> How can the thumbnail be retrieved instantly with `entity.thumbDataWithOption()` and when calling `entity.file` on the same entity it downloads it ? In other words, how can we get a thumbnail from a file that is not `locallyAvailable` ? Unless all thumbnails are stored on the device and can be retrieved without the original file.

Yes, thumbnails can be retrieved locally, since the OS may have a cache manager holding all thumbnails.

> I would guess that Apple make them available for any apps to avoid using too much disk space per app for caching always the same assets but i don't know...

That's not related to a photo manage plugin.
